src/Columns/MaskOperations.cpp

```diff
@@ -0,0 +1,219 @@
+
+INSTANTIATE(UInt8)
+INSTANTIATE(UInt16)
+INSTANTIATE(UInt32)
+INSTANTIATE(UInt64)
+INSTANTIATE(UInt128)
+INSTANTIATE(UInt256)
+INSTANTIATE(Int8)
+INSTANTIATE(Int16)
+INSTANTIATE(Int32)
+INSTANTIATE(Int64)
+INSTANTIATE(Int128)
+INSTANTIATE(Int256)
+INSTANTIATE(Float32)
+INSTANTIATE(Float64)
+INSTANTIATE(Decimal32)
+INSTANTIATE(Decimal64)
+INSTANTIATE(Decimal128)
+INSTANTIATE(Decimal256)
+INSTANTIATE(DateTime64)
+INSTANTIATE(char *)
+INSTANTIATE(UUID)
+
+#undef INSTANTIATE
+
+MaskInfo getMaskFromColumn(
+    const ColumnPtr & column,
+    PaddedPODArray<UInt8> & res,
+    bool inverted,
+    const PaddedPODArray<UInt8> * mask_used_in_expanding,
+    UInt8 default_value_in_expanding,
+    bool inverted_mask_used_in_expanding,
+    UInt8 null_value,
+    const PaddedPODArray<UInt8> * null_bytemap,
+    PaddedPODArray<UInt8> * nulls)
+{
+    if (const auto * col = checkAndGetColumn<ColumnNullable>(*column))
+    {
+        const PaddedPODArray<UInt8> & null_map = checkAndGetColumn<ColumnUInt8>(*col->getNullMapColumnPtr())->getData();
+        return getMaskFromColumn(col->getNestedColumnPtr(), res, inverted, mask_used_in_expanding, default_value_in_expanding, inverted_mask_used_in_expanding, null_value, &null_map, nulls);
+    }
+
+    size_t size = column->size();
+
+    bool is_full_column = true;
+    if (mask_used_in_expanding && mask_used_in_expanding->size() != column->size())
+    {
+        is_full_column = false;
+        size = mask_used_in_expanding->size();
+    }
+    res.resize(size);
+
+    bool only_null = column->onlyNull();
+
+    /// Some columns doesn't implement getBool() method and we cannot
+    /// convert them to mask, throw an exception in this case.
+    try
+    {
+        MaskInfo info;
+        bool value;
+        size_t column_index = 0;
+        for (size_t i = 0; i != size; ++i)
+        {
+            bool use_value_from_expanding_mask = mask_used_in_expanding && (!(*mask_used_in_expanding)[i] ^ inverted_mask_used_in_expanding);
+            if (use_value_from_expanding_mask)
+                value = inverted ? !default_value_in_expanding : default_value_in_expanding;
+            else if (only_null || (null_bytemap && (*null_bytemap)[column_index]))
+            {
+                value = inverted ? !null_value : null_value;
+                if (nulls)
+                    (*nulls)[i] = 1;
+            }
+            else
+                value = inverted ? !column->getBool(column_index) : column->getBool(column_index);
+
+            if (value)
+                info.has_once = true;
+            else
+                info.has_zeros = true;
+
+            if (is_full_column || !use_value_from_expanding_mask)
+                ++column_index;
+
+            res[i] = value;
+        }
+
+        return info;
+    }
+    catch (...)
+    {
+        throw Exception("Cannot convert column " + column.get()->getName() + " to mask", ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+    }
```

Comment:
It is bad catch every exception and swallow it.
Catch only DB::Exception, add message to it, and rethrow it.